File: playground/clip/clip.rb
require 'fileutils'

class Clip
  attr_reader :file, :like

  def initialize(dir, file, like)
    @dir = dir
    @file = file
    @like = like
    @kind = dir
  end

  # クリップ移動メソッド(安全のため、移動先を保持するだけ)
  # commitメソッドによってファイルの移動を行う。
  # 第二引数に:forceを指定すると、ファイルの移動を行う。
  def move!(dest, option=nil)
    @kind = dest

    if option == :force
      commit
    end
  end

  def commit
    if @dir == @kind
      return
    end
    
    unless Dir.exist?(@kind)
      FileUtils.mkdir_p(@kind)
    end

    FileUtils.move("#{@dir}/#{@file}", @kind)
  end

  def src_path
    "#{@dir}/#{@file}"
  end

  def dest_path
    "#{@kind}/#{@file}"
  end

end

# 文字列表現
class Clip
  def to_s
    base = "Clip(src: #{src_path}, like: #{@like})"
    if @dir != @kind
      base << " -> #{dest_path}"
    end

    base
  end
end

# クラスメソッド
class Clip
  def self.clips(src, result)
    Dir.each_child(src).map { |file| Clip.new(src, file, result[file]) }
  end
end

File: clipranking/clip.py
from dataclasses import dataclass
from typing import Optional, Iterator

from .storage import storage

@dataclass
class Clip:
    dir: str
    file: str
    like: int = 0
    dest: Optional[str] = None

    @property
    def src_path(self) -> str:
        return f"{self.dir}/{self.file}"

    @property
    def dest_path(self) -> str:
        if self.dest is None:
            return self.src_path

        return f"{self.dest}/{self.file}"

    # TODO async の使い方
    async def commit(self) -> None:
        if self.dir == self.dest:
            return
        
        if not storage.exists(self.dest):
            storage.makedirs(self.dest)
        
        storage.move(self.src_path, self.dest_path)

    def __str__(self):
        base = f"Clip(src: {self.src_path}, like: {self.like})"
        if self.dest is not None:
            base += f" -> {self.dest_path}"

        return base

    def __eq__(self, other: 'Clip') -> bool:
        return self.src_path == other.src_path \
               and self.dest_path == other.dest_path \
               and self.like == other.like

    @staticmethod
    def clips(dir, result: Optional[dict[str, int]]=None) -> Iterator['Clip']:
        if result is None:
            result = {}

        return (
            Clip(dir, file, result.get(file, 0))
            for file
            in storage.glob(dir, "*")
        )